Panic during DB writeout (invalid IPv4 / IPv6 key)

When building `goProbe` from branch `43-global-query` (commit `6821b8b414f4afea98a6bd84ef1eb0576249724a`) it crashed during (first) writeout:
```
panic: key is neither ipv4 nor ipv6

goroutine 38 [running]:
github.com/els0r/goProbe/pkg/types.Key.IsIPv4(...)
        /home/fako/Develop/go/src/github.com/els0r/goProbe/pkg/types/keyval.go:106
github.com/els0r/goProbe/pkg/types.Key.GetSip(...)
        /home/fako/Develop/go/src/github.com/els0r/goProbe/pkg/types/keyval.go:164
github.com/els0r/goProbe/pkg/types/hashmap.List.Sort.func1(0xc0001c2000?, 0xc000212f00?)
        /home/fako/Develop/go/src/github.com/els0r/goProbe/pkg/types/hashmap/list.go:43 +0x42b
sort.order2_func(...)
        /usr/local/go/src/sort/zsortfunc.go:299
sort.median_func({0xc0001432e8?, 0xc000212f00?}, 0x9e, 0x9f, 0xa0, 0xc000143188)
        /usr/local/go/src/sort/zsortfunc.go:308 +0x4a
sort.medianAdjacent_func(...)
        /usr/local/go/src/sort/zsortfunc.go:316
sort.choosePivot_func({0xc0001432e8?, 0xc000212f00?}, 0x0, 0x3c?)
        /usr/local/go/src/sort/zsortfunc.go:281 +0x110
sort.pdqsort_func({0xc0001432e8?, 0xc000212f00?}, 0x7f12e76065b8?, 0x18?, 0xc0001c2000?)
        /usr/local/go/src/sort/zsortfunc.go:89 +0xdb
sort.Slice({0x7eb240?, 0xc00021e330?}, 0xd5?)
        /usr/local/go/src/sort/slice.go:26 +0xfa
github.com/els0r/goProbe/pkg/types/hashmap.List.Sort({0xc0000bc000, 0xd5, 0xd5})
        /home/fako/Develop/go/src/github.com/els0r/goProbe/pkg/types/hashmap/list.go:39 +0x76
github.com/els0r/goProbe/pkg/goDB.dbData({0xc0001a61e0?, 0xb82a48?}, 0xc0001437a0?, 0x44dc35?)
        /home/fako/Develop/go/src/github.com/els0r/goProbe/pkg/goDB/db_writer.go:104 +0xd3
github.com/els0r/goProbe/pkg/goDB.(*DBWriter).Write(0xc000212ae0, 0xc000143d60?, {0xc00011f2f0?}, 0x4?)
        /home/fako/Develop/go/src/github.com/els0r/goProbe/pkg/goDB/db_writer.go:53 +0x23d
github.com/els0r/goProbe/pkg/goprobe/writeout.(*Handler).HandleWriteouts.func1()
        /home/fako/Develop/go/src/github.com/els0r/goProbe/pkg/goprobe/writeout/handler.go:217 +0x74d
created by github.com/els0r/goProbe/pkg/goprobe/writeout.(*Handler).HandleWriteouts
        /home/fako/Develop/go/src/github.com/els0r/goProbe/pkg/goprobe/writeout/handler.go:179 +0xaa
```

It's unclear where that comes from (and there's been a lot of changes to slimcap that are not yet merged / matched in `goProbe`), but still this should be investigated.

In addition, upon restarting `goProbe` without removing the DB directory there was an error upon next (and all consecutive) writeouts:
```
2023-04-05T11:30:00.091+0200    error   writeout/handler.go:222 Error during writeout: seek /tmp/goprobe_test/eth0/2023/04/1680652800/proto.gpf: invalid argument       {"app_name": "goProbe_alpine", "app_version": "872abf81"}
```

This indicates that the failure protection implemented in #60 doesn't work as intended (maybe because it failed on the first attempt, leaving an empty file that now can't be `seek`ed:
```
fw-2:~# ll /tmp/goprobe_test/eth0/2023/04/1680652800/proto.gpf
-rw-r--r--    1 root     root             0 Apr  5 11:30 /tmp/goprobe_test/eth0/2023/04/1680652800/proto.gpf
```

DoD

- [x] Reproduce write failure on first attempt (and add case to tests) and handle accordingly
- [x] Track potential issue with invalid keys (as the panic indicates, this should not happen (tm))
- [x] Fix discovered hashmap iterator issue

First analysis shows that it wants to seek to, well, somewhere that doesn't exist for sure:
```
2023-04-05T12:10:48.575+0200    error   writeout/handler.go:222 Error during writeout: seek to 72057594037927936 failed: seek /tmp/goprobe_test/eth0/2023/04/1680652800/proto.gpf: invalid argument    {"app_name": "goProbe_alpine_debug", "app_version": "872abf81"}
```
In addition, the file size of the `dip` column is huge, and a metadata analysis shows utter garbage:
```
└─ $ ▶ ./analyze-meta -path /tmp/goprobe_test/eth0/2023/04/1680652800/.blockmeta 

              Column: sip
    Number of Blocks: 0
     Number of Flows: 0 / 0 (IP v4 / v6)
             Traffic: bytes: received=0 sent=0; packets: received=0 sent=0
                Size: 4 bytes

        #    ts    time UTC    offset    len    raw len    encoder    ratio    #F v4    #F v6    drops    first 4 bytes        integrity check
                                                                                                                                                  
              Column: dip
    Number of Blocks: 0
     Number of Flows: 0 / 0 (IP v4 / v6)
             Traffic: bytes: received=0 sent=0; packets: received=0 sent=0
                Size: 17179869188 bytes

        #    ts    time UTC    offset    len    raw len    encoder    ratio    #F v4    #F v6    drops    first 4 bytes        integrity check
                                                                                                                                                  
              Column: proto
    Number of Blocks: 0
     Number of Flows: 0 / 0 (IP v4 / v6)
             Traffic: bytes: received=0 sent=0; packets: received=0 sent=0
                Size: 72057594037927936 bytes

        #    ts    time UTC    offset    len    raw len    encoder    ratio    #F v4    #F v6    drops    first 4 bytes        integrity check
                                                                                                                                                  
              Column: dport
    Number of Blocks: 0
     Number of Flows: 0 / 0 (IP v4 / v6)
             Traffic: bytes: received=0 sent=0; packets: received=0 sent=0
                Size: 288230376218820608 bytes

        #    ts    time UTC    offset    len    raw len    encoder    ratio    #F v4    #F v6    drops    first 4 bytes        integrity check
                                                                                                                                                  
              Column: bytes_rcvd
    Number of Blocks: 0
     Number of Flows: 0 / 0 (IP v4 / v6)
             Traffic: bytes: received=0 sent=0; packets: received=0 sent=0
                Size: 288511851128422400 bytes

        #    ts    time UTC    offset    len    raw len    encoder    ratio    #F v4    #F v6    drops    first 4 bytes        integrity check
                                                                                                                                                  
              Column: bytes_sent
    Number of Blocks: 0
     Number of Flows: 0 / 0 (IP v4 / v6)
             Traffic: bytes: received=0 sent=0; packets: received=0 sent=0
                Size: 281474976776192 bytes

        #    ts    time UTC    offset    len    raw len    encoder    ratio    #F v4    #F v6    drops    first 4 bytes        integrity check
                                                                                                                                                  
              Column: pkts_rcvd
    Number of Blocks: 0
     Number of Flows: 0 / 0 (IP v4 / v6)
             Traffic: bytes: received=0 sent=0; packets: received=0 sent=0
                Size: 282574488338432 bytes

        #    ts    time UTC    offset    len    raw len    encoder    ratio    #F v4    #F v6    drops    first 4 bytes        integrity check
                                                                                                            
              Column: pkts_sent
    Number of Blocks: 0
     Number of Flows: 0 / 0 (IP v4 / v6)
             Traffic: bytes: received=0 sent=0; packets: received=0 sent=0
                Size: 2199023256064 bytes

        #    ts    time UTC    offset    len    raw len    encoder    ratio    #F v4    #F v6    drops    first 4 bytes        integrity check
```
Almost seems as if the sizes are more or less random, as if uninitialized. If I wanted that I could have used C... :roll_eyes:

I'm able to semi-reliably reproduce the issue on one of my sensors, and it seems that the issue is an empty key:
```
panic: key `[]` is neither ipv4 nor ipv6
```
I'll add some more debugging statements to figure out where that comes from.

--- a/examples/analyze-meta/main.go
+++ b/examples/analyze-meta/main.go
@@ -41,7 +41,7 @@ func main() {
 	if err != nil {
 		logger.Fatalf("failed to extract timestamp: %s", err)
 	}
-	baseDirPath := filepath.Dir(dirPath)
+	baseDirPath := filepath.Dir(filepath.Dir(filepath.Dir(dirPath)))
 
 	gpDir := gpfile.NewDir(baseDirPath, timestamp, gpfile.ModeRead)
 	if err := gpDir.Open(); err != nil {

--- a/pkg/goDB/db_writer.go
+++ b/pkg/goDB/db_writer.go
@@ -48,7 +48,6 @@ func (w *DBWriter) Write(flowmap *hashmap.AggFlowMap, captureMeta CaptureMetadat
 		err = fmt.Errorf("Could not create / open daily directory: %w", err)
 		return err
 	}
-	defer dir.Close()
 
 	data, update = dbData(w.iface, timestamp, flowmap)
 	if err := dir.WriteBlocks(timestamp, gpfile.TrafficMetadata{
@@ -59,9 +58,10 @@ func (w *DBWriter) Write(flowmap *hashmap.AggFlowMap, captureMeta CaptureMetadat
 		return err
 	}
 
-	return err
+	return dir.Close()
 }
 
+// BulkWorkload denotes a set of workloads / writes to perform during WriteBulk()
 type BulkWorkload struct {
 	FlowMap     *hashmap.AggFlowMap
 	CaptureMeta CaptureMetadata
@@ -80,7 +80,6 @@ func (w *DBWriter) WriteBulk(workloads []BulkWorkload, dirTimestamp int64) (err 
 		err = fmt.Errorf("Could not create / open daily directory: %w", err)
 		return err
 	}
-	defer dir.Close()
 
 	for _, workload := range workloads {
 		data, update = dbData(w.iface, workload.Timestamp, workload.FlowMap)
@@ -93,7 +92,7 @@ func (w *DBWriter) WriteBulk(workloads []BulkWorkload, dirTimestamp int64) (err 
 		}
 	}
 
-	return nil
+	return dir.Close()
 }
 
 func dbData(iface string, timestamp int64, aggFlowMap *hashmap.AggFlowMap) ([types.ColIdxCount][]byte, gpfile.Stats) {

--- a/pkg/goDB/db_writer_test.go
+++ b/pkg/goDB/db_writer_test.go
@@ -16,7 +16,7 @@ import (
 
 func TestPanicDuringWrite(t *testing.T) {
 
-	// Create a temporary file (in the destinantion directory to avoid moving accross the FS barrier)
+	// Setup a temporary directory for the test DB
 	tempDir, err := os.MkdirTemp(os.TempDir(), "dbwrite_test")
 	require.Nil(t, err)
 	defer require.Nil(t, os.RemoveAll(tempDir))

--- a/pkg/goDB/db_writer_test.go
+++ b/pkg/goDB/db_writer_test.go
@@ -1,0 +1,61 @@
+package goDB
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/els0r/goProbe/pkg/goDB/encoder/encoders"
+	"github.com/els0r/goProbe/pkg/goDB/storage/gpfile"
+	"github.com/els0r/goProbe/pkg/types/hashmap"
+	"github.com/stretchr/testify/require"
+)
+
+func TestPanicDuringWrite(t *testing.T) {
+
+	// Create a temporary file (in the destinantion directory to avoid moving accross the FS barrier)
+	tempDir, err := os.MkdirTemp(os.TempDir(), "dbwrite_test")
+	require.Nil(t, err)
+	defer require.Nil(t, os.RemoveAll(tempDir))
+
+	timestamp := time.Now().Unix()
+	dayTimestamp := gpfile.DirTimestamp(timestamp)
+	dayUnix := time.Unix(dayTimestamp, 0)
+	dirPath := filepath.Join(filepath.Join(tempDir, "test"), strconv.Itoa(dayUnix.Year()), fmt.Sprintf("%02d", dayUnix.Month()), strconv.Itoa(int(dayTimestamp)))
+
+	w := NewDBWriter(tempDir, "test", encoders.EncoderTypeNull)
+
+	// Add a single item that will trigger a panic later
+	testMap := hashmap.NewAggFlowMap()
+	testMap.V4Map.Set([]byte{0x0}, hashmap.Val{})
+
+	t.Run("Write", func(t *testing.T) {
+		require.Panics(t, func() {
+			err := w.Write(testMap, CaptureMetadata{}, timestamp)
+			_ = err
+		})
+		dirs, err := os.ReadDir(dirPath)
+		require.Nil(t, err)
+		require.Empty(t, dirs)
+	})
+
+	t.Run("WriteBulk", func(t *testing.T) {
+		require.Panics(t, func() {
+			err := w.WriteBulk([]BulkWorkload{
+				{
+					FlowMap:     testMap,
+					CaptureMeta: CaptureMetadata{},
+					Timestamp:   timestamp,
+				},
+			}, timestamp)
+			_ = err
+		})
+		dirs, err := os.ReadDir(dirPath)
+		require.Nil(t, err)
+		require.Empty(t, dirs)
+	})
+
+}

--- a/pkg/goDB/storage/gpfile/gpdir.go
+++ b/pkg/goDB/storage/gpfile/gpdir.go
@@ -88,7 +88,8 @@ func newMetadata() *Metadata {
 	}
 	for i := 0; i < int(types.ColIdxCount); i++ {
 		m.BlockMetadata[i] = &storage.BlockHeader{
-			BlockList: make([]storage.BlockAtTime, 0),
+			CurrentOffset: 0,
+			BlockList:     make([]storage.BlockAtTime, 0),
 		}
 	}
 	return &m

--- a/pkg/goDB/storage/gpfile/gpfile.go
+++ b/pkg/goDB/storage/gpfile/gpfile.go
@@ -304,7 +304,7 @@ func (g *GPFile) open(flags int) (err error) {
 		// Ensure that the file is loaded at the position of the last known successful write
 		// The bufio.Writer will honor that position, even after a Reset()
 		if _, err = g.file.Seek(g.header.CurrentOffset, 0); err != nil {
-			return
+			return fmt.Errorf("seek to %d failed: %w", g.header.CurrentOffset, err)
 		}
 		g.fileWriteBuffer = bufio.NewWriter(g.file)
 	}

--- a/pkg/types/hashmap/hashmap_test.go
+++ b/pkg/types/hashmap/hashmap_test.go
@@ -79,6 +79,42 @@ func TestHashMapSetOrUpdate(t *testing.T) {
 	require.Equal(t, count, testMap.Len())
 }
 
+func TestHashMapIteratorConsistency(t *testing.T) {
+	testMap := New()
+	for i := 0; i < 1000; i++ {
+		temp := make([]byte, 8)
+		binary.BigEndian.PutUint64(temp, uint64(i))
+		testMap.Set(temp, types.Counters{BytesRcvd: uint64(i), BytesSent: 0, PacketsRcvd: 0, PacketsSent: 0})
+
+		var count int
+		for it := testMap.Iter(); it.Next(); {
+			count++
+		}
+		require.Equal(t, i+1, testMap.Len())
+		require.Equal(t, testMap.Len(), count)
+	}
+}
+
+func TestAggFlowMapFlatten(t *testing.T) {
+	testMap := NewAggFlowMap()
+	for i := 0; i < 1000; i++ {
+		temp := make([]byte, 8)
+		binary.BigEndian.PutUint64(temp, uint64(i))
+		testMap.V4Map.Set(temp, types.Counters{BytesRcvd: uint64(i), BytesSent: 0, PacketsRcvd: 0, PacketsSent: 0})
+		testMap.V6Map.Set(temp, types.Counters{BytesRcvd: uint64(i), BytesSent: 0, PacketsRcvd: 0, PacketsSent: 0})
+		var count int
+		for it := testMap.Iter(); it.Next(); {
+			count++
+		}
+		require.Equal(t, 2*(i+1), testMap.Len())
+		require.Equal(t, testMap.Len(), count)
+
+		v4List, v6List := testMap.Flatten()
+		require.Equal(t, i+1, len(v4List))
+		require.Equal(t, i+1, len(v6List))
+	}
+}
+
 func TestLinearHashMapOperations(t *testing.T) {
 
 	testMap := New()

--- a/pkg/types/hashmap/iterator.go
+++ b/pkg/types/hashmap/iterator.go
@@ -1,6 +1,8 @@
 package hashmap
 
-import "github.com/zeebo/xxh3"
+import (
+	"github.com/zeebo/xxh3"
+)
 
 // Iter provides a map Iter to allow traversal
 type Iter struct {
@@ -76,7 +78,7 @@ next:
 		}
 		k := b.keys[offi]
 		if checkBucket != noBucket && !m.sameSizeGrow() {
-			hash := xxh3.Hash(k)
+			hash := xxh3.HashSeed(k, m.seed)
 			if int(hash&m.bucketMask()) != checkBucket {
 				continue
 			}

--- a/pkg/types/keyval.go
+++ b/pkg/types/keyval.go
@@ -103,7 +103,7 @@ func (k Key) IsIPv4() bool {
 	if len(k) == KeyWidthIPv6 {
 		return false
 	}
-	panic("key is neither ipv4 nor ipv6")
+	panic(fmt.Sprintf("key `%v` is neither ipv4 nor ipv6", []byte(k)))
 }
 
 // Len returns the length of the key (e.g. to determine the IP version)
@@ -230,7 +230,7 @@ func (e ExtendedKey) IsIPv4() bool {
 	if len(e) == KeyWidthIPv6 || len(e) == KeyWidthIPv6+TimestampWidth {
 		return false
 	}
-	panic("extended key is neither ipv4 nor ipv6")
+	panic(fmt.Sprintf("extended key `%v` is neither ipv4 nor ipv6", []byte(e)))
 }
 
 // PutSip stores a source IP in the key

